ambry-network/src/main/java/com/github/ambry/network/FifoNetworkRequestQueue.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright 2022 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.network;
+
+import com.github.ambry.utils.Time;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class FifoNetworkRequestQueue implements NetworkRequestQueue {
+  private static final Logger logger = LoggerFactory.getLogger(FifoNetworkRequestQueue.class);
+  private final int timeout;
+  private final Time time;
+  private final BlockingQueue<NetworkRequest> queue;
+
+  FifoNetworkRequestQueue(int capacity, int timeout, Time time) {
+    this.timeout = timeout;
+    this.time = time;
+    queue = new ArrayBlockingQueue<>(capacity);
+  }
+
+  @Override
+  public boolean offer(NetworkRequest request) {
+    return queue.offer(request);
+  }
+
+  @Override
+  public NetworkRequestBundle take() throws InterruptedException {
+    NetworkRequest requestToServe = null;
+    List<NetworkRequest> requestsToDrop = new ArrayList<>();
+    NetworkRequest nextRequest;
+    while ((nextRequest = queue.poll()) != null) {
+      if (needToDrop(nextRequest)) {
+        logger.warn("Request timed out waiting in queue, dropping it: {}", nextRequest);
+        requestsToDrop.add(nextRequest);
```

Comment:
When "needToDrop" is true, shall we also need return a request to server?
Looks like we either return one request to serve or one request to drop.

But for the definition of NetworkRequestBundle, it can have 1 request to serve and multiple requests to drop.

Reply:
Hmm, in this case, if we find that the request needs to be dropped, we are not actually breaking from loop and are continuing to iterate until we find a valid request or the queue is empty. So, the `NetworkRequestBundle` could return 1  request to serve and multiple requests to drop.